rest-service/manager_rest/storage/resource_models.py

```diff
@@ -2024,6 +2093,70 @@ def target_deployment(cls):
     source_deployment_id = association_proxy('source_deployment', 'id')
     target_deployment_id = association_proxy('target_deployment', 'id')
 
+    @classmethod
+    def _dependencies_adjacency(cls, deployment_ids, dependents=True):
+        """Select a dependency subgraph in an adjacency-list form.
+
+        Returns a query yielding (idd_id, source_id, target_id), which mean
+        "IDD with id idd_id, says that deployment source_id depends on the
+        deployment target_id".
+
+        This is recursive, so will return parents, then parents of parents,
+        etc. (or children, then children of children, etc).
+
+        :param deployment_ids: storage_ids of the root deployments
+        :param dependents: if set, return dependents, ie. children; otherwise,
+            return dependencies, ie. parents
+        """
+        select_cols = db.session.query(
+            cls._storage_id,
+            cls._source_deployment,
+            cls._target_deployment,
+        )
+
+        if dependents:
+            base = select_cols.filter(
+                cls._target_deployment.in_(deployment_ids))
+        else:
+            base = select_cols.filter(
+                cls._source_deployment.in_(deployment_ids))
+        base = base.cte(name='dependents', recursive=True)
```

Comment:
for a recap of recursive CTEs, check out https://www.postgresql.org/docs/14/queries-with.html